omniduct/caches/base.py

```diff
@@ -99,26 +96,115 @@ def __init__(self, **kwargs):
     def _init(self):
         pass
 
-    @abstractmethod
-    def clear(self, id_duct, id_str):
+    # Data insertion and retrieval
+
+    def set(self, key, value, namespace=None, serializer=PickleSerializer, expires=None, metadata=None):
+        namespace = self._namespace(namespace)
+        key = self._key(key)
+        # try:
+        self.set_metadata(key, metadata, namespace=namespace, replace=True)
+        with self._get_stream_for_key(namespace, key, 'data{}'.format(serializer.file_extension), mode='wb', create=True) as fh:
+            return serializer.serialize(value, fh)
+        # except:
+        #     self.unset(key, namespace=namespace)
+
+    def set_metadata(self, key, metadata, namespace=None, replace=False):
+        namespace = self._namespace(namespace)
+        key = self._key(key)
+        if replace:
+            orig_metadata = {'created': datetime.datetime.utcnow()}
+        else:
+            orig_metadata = self.get_metadata(key, namespace=namespace)
+
+        orig_metadata.update(metadata or {})
+
+        with self._get_stream_for_key(namespace, key, 'metadata', mode='w', create=True) as fh:
+            yaml.safe_dump(orig_metadata, fh, default_flow_style=False)
+
+    def get(self, key, namespace=None, serializer=PickleSerializer):
+        namespace = self._namespace(namespace)
+        key = self._key(key)
+        try:
+            with self._get_stream_for_key(namespace, key, 'data{}'.format(serializer.file_extension), mode='rb', create=False) as fh:
+                return serializer.deserialize(fh)
+        finally:
+            self.set_metadata(key, namespace=namespace, metadata={'last_accessed': datetime.datetime.utcnow()})
```

Comment:
hm ... I like the metadata tracking but worry a bit about performance, perhaps this should happen through an instance flag

Reply:
For local filesystems, and in my experience, performance is not an issue. I agree that if you are using a remote filesystem, performance could get painful.

In future releases, this metadata will be used to track and control resource usage, data expiry, and so on, so having this metadata is useful. Disabling it for some filesystems is eminently doable, but would not require an API break, and so perhaps we can just see if people use it on remote / slow filesystems, and if so, do something about it?